src/classes/ADDR_Validator_TDTM.cls

```diff
@@ -37,59 +37,65 @@
 */
 public class ADDR_Validator_TDTM extends TDTM_Runnable {
-	
-	/*******************************************************************************************************
+
+    /*******************************************************************************************************
     * @description Exception raised during the address validation process.
     */
-	public class ADDR_Validator_Exception extends Exception { }
-	
-	/*******************************************************************************************************
-    * @description Indicates whether the class has been run in the current context.
-    */
-	public static Boolean alreadyRun = false;
-	
-	/*******************************************************************************************************
+    public class ADDR_Validator_Exception extends Exception { }
+
+    /*******************************************************************************************************
     * @description Verifies new or modified address records, if verification enabled in the settings.
     * @param listNew The list of Addresses from trigger new. 
     * @param listOld The list of Addresses from trigger old. 
     * @param triggerAction The trigger event (BeforeInsert, AfterInsert, etc.) that caused the class to run. 
     * @param objResult The describe for Addresses.
     * @return dmlWrapper The wrapper with the records to perform DML on. 
     */
-	public override DmlWrapper run(List<SObject> newList, List<SObject> oldList, 
-    TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) { 
+    public override DmlWrapper run(List<SObject> newList, List<SObject> oldList,
+    TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+
+		// Look in the static list of previoysly processed Id's to determine if the record has already
+		// gone through this trigger. If so, then don't process it a second time.
+        Boolean recordsPreviouslyProcessed = false;
+        if (newList[0].Id != null && ADDR_ProcessControl.processedIdsForValidation.contains(newList[0].Id)) {
+			recordsPreviouslyProcessed = true;
+        }
         
-        if(!alreadyRun) {
-	        Addr_Verification_Settings__c settings = UTIL_CustomSettingsFacade.getAddressVerificationSettings();
-	        if ((triggerAction == TDTM_Runnable.Action.BeforeInsert || triggerAction == TDTM_Runnable.Action.BeforeUpdate) &&
-	           settings.enable_automatic_verification__c) {        	
-	        	for(SObject so : newList) {
-	        		Address__c address = (Address__c)so;
-	        		if(!system.isFuture() && !system.isBatch()) {
-	        		    address.Verification_Status__c = Label.Addr_Pending_Verification;
-	        		} else {
-	        			address.Verification_Status__c = Label.Addr_Verification_Required;
-	        		}	
-        	    }
-	        }
+        if(!ADDR_ProcessControl.checkValidationRecursionState() && !recordsPreviouslyProcessed) {
```

Comment:
I thought of it this way:
- The initial call to the class saved 200 of Id's being validated.
- A subsequent recursive call would re-submit one or all of those 200 records. 
- Thus if the first Id in the triggerNew collection exists in the saved list of Id's, then it's recursive and the whole batch can be ignored.

Though, I suppose there is another possible path:
- Recursion could submit a combination of previously processed Addresses _and_ one or more new addresses for validation.
- I'm not necessarily sure that is possible though given that we're controlling the address recursion. 
- Though unlikely, I suppose that a org could have a custom or managed trigger that inserts a new Address after a previously inserted/updated Address is validated. Certainly an outlier, but probably not impossible.